lib/pos_snapshot_entries.go

```diff
@@ -0,0 +1,764 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, "DBGetSnapshotGlobalActiveStakeAmountNanosWithTxn: problem decoding value")
+	}
+	return globalActiveStakeAmountNanos, nil
+}
+
+func DBPutSnapshotGlobalActiveStakeAmountNanosWithTxn(
+	txn *badger.Txn,
+	snap *Snapshot,
+	globalActiveStakeAmountNanos *uint256.Int,
+	snapshotAtEpochNumber uint64,
+	blockHeight uint64,
+) error {
+	if globalActiveStakeAmountNanos == nil {
+		// This should never happen but is a sanity check.
+		glog.Errorf("DBPutSnapshotGlobalActiveStakeAmountNanos: called with nil GlobalActiveStake, this should never happen")
+		return nil
+	}
+	key := DBKeyForSnapshotGlobalActiveStakeAmountNanos(snapshotAtEpochNumber)
+	return DBSetWithTxn(txn, snap, key, VariableEncodeUint256(globalActiveStakeAmountNanos))
+}
+
+//
+// SnapshotLeaderScheduleValidator
+//
+
+type SnapshotLeaderScheduleMapKey struct {
+	SnapshotAtEpochNumber uint64
+	LeaderIndex           uint16
+}
+
+func (bav *UtxoView) GetSnapshotLeaderScheduleValidator(leaderIndex uint16) (*ValidatorEntry, error) {
+	// Calculate the SnapshotEpochNumber.
+	snapshotAtEpochNumber, err := bav.GetSnapshotEpochNumber()
+	if err != nil {
+		return nil, errors.Wrapf(err, " GetSnapshotLeaderScheduleValidator: problem calculating SnapshotEpochNumber: ")
+	}
+	// First, check the UtxoView.
+	mapKey := SnapshotLeaderScheduleMapKey{SnapshotAtEpochNumber: snapshotAtEpochNumber, LeaderIndex: leaderIndex}
+	if validatorPKID, exists := bav.SnapshotLeaderSchedule[mapKey]; exists {
+		return bav.GetSnapshotValidatorByPKID(validatorPKID)
+	}
+	// Next, check the db.
+	validatorEntry, err := DBGetSnapshotLeaderScheduleValidator(bav.Handle, bav.Snapshot, leaderIndex, snapshotAtEpochNumber)
+	if err != nil {
+		return nil, errors.Wrapf(err, "GetSnapshotLeaderScheduleValidator: error retrieving ValidatorPKID: ")
+	}
+	if validatorEntry != nil {
+		// Cache the ValidatorPKID in the UtxoView.
+		bav._setSnapshotLeaderScheduleValidator(validatorEntry.ValidatorPKID, leaderIndex, snapshotAtEpochNumber)
+	}
+	return validatorEntry, nil
+}
+
+func (bav *UtxoView) _setSnapshotLeaderScheduleValidator(validatorPKID *PKID, index uint16, snapshotAtEpochNumber uint64) {
+	if validatorPKID == nil {
+		glog.Errorf("_setSnapshotLeaderScheduleValidator: called with nil ValidatorPKID, this should never happen")
+		return
+	}
+	mapKey := SnapshotLeaderScheduleMapKey{SnapshotAtEpochNumber: snapshotAtEpochNumber, LeaderIndex: index}
+	bav.SnapshotLeaderSchedule[mapKey] = validatorPKID.NewPKID()
+}
+
+func (bav *UtxoView) _flushSnapshotLeaderScheduleToDbWithTxn(txn *badger.Txn, blockHeight uint64) error {
+	for mapKey, validatorPKID := range bav.SnapshotLeaderSchedule {
+		if validatorPKID == nil {
+			return fmt.Errorf(
+				"_flushSnapshotLeaderScheduleToDb: found nil PKID for EpochNumber %d, this should never happen",
+				mapKey.SnapshotAtEpochNumber,
+			)
+		}
+		if err := DBPutSnapshotLeaderScheduleValidatorWithTxn(
```

Comment:
Here again I would follow the pattern of DELETE -> PUT

Reply:
The key-value here is SnapshotAtEpochNumber uint64, LeaderIndex uint16 --> ValidatorPKID *PKID. IMO there isn't a convenient place to store the isDeleted field. IMO we shouldn't delete the leader schedule rows, only overwrite. Also you may be suggesting we do an O(n^2) operation to make sure the ValidatorPKID doesn't show up elsewhere in the leader schedule when we set it (otherwise delete it), but IMO not worth it.

The only edge case that could happen here is if we update the leader schedule length and one is 100, the next is 99. There would be one validator left-over from the first leader schedule. But 1) we shouldn't ever be overwriting a leader schedule once set, and 2) a change to the "how many validators do we put in a leader schedule" param wouldn't happen within the same epoch.